### PR TITLE
First step at container aliasing

### PR DESCRIPTION
**Pull request by eellison**
First step at allowing container types within alias analysis.

Since the current implementation hides the concept of Wildcards within alias analysis and does not expose it to memory dag, we cannot represent whether a container type holds a wildcard. As a result, only handle TupleConstruct, where we can directly inspect if any input values are wildcards, and don't handle nested containers.

### CODE REVIEW

**Comment by eellison** [comment]
Landing this is a strict blocker on first class functions, FYI @suo, as well as a hard-to-hit bug that exists currently

File: torch/csrc/jit/passes/utils/memory_dag.h [hunk 1]
```diff
@@ -38,6 +41,10 @@ class MemoryDAG {
   bool mayAlias(const Element* a, const Element* b) const;
   bool mayAlias(Element* a, Element* b) const;
 
+  // Does container hold reference to any memory that is stored in elem ?
+  bool mayContainAlias(const Element* elem, const Element* container) const;
```
Comment (suo) [inline torch/csrc/jit/passes/utils/memory_dag.h]: The implementation of this is commutative but the argument names and comment make it seem like it's not. Can we change it to `mayContainAlias(a, b)` and update the comment to make it clear that the order of the arguments does not matter?

File: torch/csrc/jit/passes/alias_analysis.h [hunk 2]
```diff
@@ -45,6 +45,9 @@ class AliasDb {
   bool writesToAlias(Node* n, const ValueSet& vs, bool recurseBlocks = false)
       const;
 
+  // Does `container` hold in memory any element that exist in elem
+  bool mayContainAlias(const Value* elem, const Value* container) const;
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.h]: This implementation is not commutative but the one in MemoryDAG is. This is potentially pretty confusing, and I see no blocker to making both implementations commutative

**Review by suo — CHANGES_REQUESTED** [review]
This approach will unfortunately not work for any other containers, but I guess we should unbreak the CSE bug for now.

File: torch/csrc/jit/passes/alias_analysis.h (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.h]: i guess i would just call the current function with the order of the arguments reversed

**Comment by eellison** [comment]
> This approach will unfortunately not work for any other containers, but I guess we should unbreak the CSE bug for now.

As far as the memory dag I think it would still continue to.

**Review by eellison — COMMENTED** [review]
Made a few small changes. 

@suo The current implementation checks in memory dag that `b` contains `a` but not that `a` contains `b`. Is that what we want? It might simplify things things to make it commutative.

Edit: recent commit makes it commutative

**Comment by eellison** [comment]
@suo take another look please ?

File: torch/csrc/jit/passes/utils/memory_dag.cpp [hunk 3]
```diff
@@ -31,12 +32,65 @@ bool MemoryDAG::mayAliasImpl(const Element* a, const Element* b) const {
   return false;
 }
 
+bool MemoryDAG::mayContainAlias(
```
Comment (suo) [inline torch/csrc/jit/passes/utils/memory_dag.cpp]: Let's make the argument names more consistent (here and in the header file). In particular, I think "contained"/"container" and "elem/container" are misleading with the new commutative implementation.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 4]
```diff
@@ -772,6 +810,57 @@ bool AliasDb::mayAlias(const Value* a, const Value* b) const {
...
+    return false;
+  }
+
+  // don't know how to analyze containers other than tuple
+  if (isContainerType(elem->type())) {
+    if (elem->node()->kind() != prim::TupleConstruct) {
+      return true;
+    }
+    auto inp = elem->node()->inputs();
+    return std::any_of(inp.begin(), inp.end(), [&](const Value* v) {
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.cpp]: I don't understand why you're comparing aliasing on the contained types directly. Shouldn't `MemoryDAG` do that for you?

The role of AliasDb wrappers are to do wildcard/shouldAnnotate processing on the values before dispatching to MemoryDAG for the actual query. So in this case, I think you should recursively check for wildcards in the lhs/rhs containers, then dispatch directly to `MemoryDag::mayContainAlias` to do the rest.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 5]
```diff
@@ -772,6 +810,57 @@ bool AliasDb::mayAlias(const Value* a, const Value* b) const {
   return memoryDAG_->mayAlias(elementMap_.at(a), elementMap_.at(b));
 }
 
+bool AliasDb::mayContainAlias(const Value* elem, const Value* container) const {
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.cpp]: nit: elem/container should get renamed.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 6]
```diff
@@ -772,6 +810,57 @@ bool AliasDb::mayAlias(const Value* a, const Value* b) const {
...
+    return false;
+  }
+
+  // don't know how to analyze anything except TupleConstruct
+  if (container->node()->kind() != prim::TupleConstruct) {
+    return true;
+  }
+
+  auto inp = container->node()->inputs();
+  if (std::any_of(inp.begin(), inp.end(), [&](const Value* cont) {
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Same here: doesn't MemoryDAG look inside the contained type for you?

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 7]
```diff
@@ -772,6 +810,57 @@ bool AliasDb::mayAlias(const Value* a, const Value* b) const {
...
+        return mayContainAlias(elem, cont);
+      })) {
+    return true;
+  };
+
+  for (const auto& input : container->node()->inputs()) {
+    if (isWildcard(input)) {
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.cpp]: You should do this check for both the lhs and rhs potentially, no?

File: torch/csrc/jit/passes/alias_analysis.cpp (hunk 4 above)
Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: You made the comment that we could make the logic recursive, that is what this is doing. Memory Dag can't do it for me because there are certain types that we have to handle at the border, including Wildcards but also container types that we don't know how to analyze yet - everything but tupleconstruct.

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: I meant that you should separate out the following two steps:
1. Special case handling: check if there are any wildcards or non-supported container types in the lhs and rhs container trees. Return true if the answer is yes.
2. If there are no special cases, dispatch the memoryDAG to answer the query normally.

Right now you're doing a variation of what you were doing before—recursive comparisons on both sides of contained value tree, with each comparison doing its own traversal of both subtrees (the memory DAG call).

Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Do you know why the 
`
  if (!elementMap_.count(a) || !elementMap_.count(b)) {
    return false;
  }
` 
check exists in mayAlias?

Comment (suo) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Not sure; if you can remove it would be good—the invariant should be that any value for which `shouldAnnotate` returns true should have an element.

Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Yep that's what i thought. will try to remove.

**Comment by eellison** [comment]
I think this is good to land

**Review by suo — APPROVED** [review]

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 8]
```diff
@@ -259,15 +269,16 @@ void AliasDb::dump() const {
...
 void AliasDb::makeAllAlias(const std::vector<Value*>& values) {
-  if (values.size() > 0) {
-    giveFreshAlias(values[0]);
+  if (values.size() == 0 || !shouldAnnotate(values[0])) {
+    return;
   }
-  for (const auto value : values) {
-    makePointerTo(value, values[0]);
+
+  giveFreshAlias(values[0]);
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.cpp]: why was this changed? This is weirder than before—now we have multiple values with the same assigned element (and the `value` field of that element only points to the first value).

Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: Thought it would optimize number of memory locations, reverted
